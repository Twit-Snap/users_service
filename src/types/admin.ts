export interface Admin {
    username: string;
    email: string;
}

export interface AdminInfoDto {
    username: string;
    email: string;
    password: string;
}

export interface LoginAdminDto {
    email: string;
    password: string;
}