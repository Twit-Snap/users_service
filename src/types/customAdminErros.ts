export class AlreadyExistError extends Error {
    entityName: string;
    entityEmail: string;
    detail: string

    constructor(entityName: string, entityEmail: string, detail: string) {

        super(`${entityName ? entityName : ''} ${entityEmail ? entityEmail : ''} already exists`);
        this.entityEmail = entityEmail;
        this.entityName = entityName;
        this.detail = detail
        this.name = 'InvalidCredentials';
    }

}

export class ValidationError extends Error {
    field: string;
    detail: string;

    constructor(field: string, detail: string) {
        super(`Validation error: ${field}`);
        this.field = field;
        this.detail = detail;
        this.name = 'ValidationError';
    }
}

export class InvalidCredentialsError extends Error {
    detail: string;

    constructor(detail: string) {

        super(`Email or password are incorrect`);
        this.detail = detail
        this.name = 'InvalidCredentials';
    }

}

export class NotFoundError extends Error {
    entityId: string;
    constructor(entityId: string) {
        super(`${entityId} not found`);
        this.entityId = entityId;
        this.name = 'NotFoundError';
    }
}