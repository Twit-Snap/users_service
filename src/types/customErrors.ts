export class NotFoundError extends Error {
  entityName: string;
  entityId: string;

  constructor(entityName: string, entityId: string) {
    super(`${entityName} not found`);
    this.entityName = entityName;
    this.entityId = entityId;
    this.name = 'NotFoundError';
  }
}

export class ValidationError extends Error {
  field: string;
  detail: string;

  constructor(field: string, detail: string) {
    super(`Validation error: ${field}`);
    this.field = field;
    this.detail = detail;
    this.name = 'ValidationError';
  }
}
