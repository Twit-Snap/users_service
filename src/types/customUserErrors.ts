export class NotFoundError extends Error {
  entityName: string;
  entityId: string;

  constructor(entityName: string, entityId: string) {
    super(`${entityName} not found`);
    this.entityName = entityName;
    this.entityId = entityId;
    this.name = 'NotFoundError';
  }
}

export class ValidationError extends Error {
  field: string;
  detail: string;
  type: string;

  constructor(field: string, detail: string, type: string) {
    super(`Validation error: ${field}`);
    this.field = field;
    this.detail = detail;
    this.name = 'ValidationError';
    this.type = type;
  }
}

export class EntityAlreadyExistsError extends Error {
  entityName: string;

  constructor(entityName: string) {
    super(`${entityName} already exists`);
    this.entityName = entityName;
    this.name = 'EntityAlreadyExistsError';
  }
}

export class AuthenticationError extends Error {

  constructor() {
    super('Authentication error');
    this.name = 'AuthenticationError';
  }
}