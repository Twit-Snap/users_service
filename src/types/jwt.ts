import { Request } from 'express';
import { JwtPayload } from 'jsonwebtoken';

// Interface definition
export interface IJWTService {
  sign(payload: JwtCustomPayload): string;
  verify(token: string): JwtPayload | string;
  decode(token: string): JwtPayload | string | null;
}

export type JwtCustomPayload = JwtUserPayload | JwtAdminPayload;

export type JwtUserPayload = {
  type: 'user';
  userId: number;
  email: string;
  username: string;
  phoneNumber: string;
};

export type JwtAdminPayload = {
  type: 'admin';
  username: string;
  email: string;
};

export type UserRequest = Request & { user: JwtUserPayload };
export type AdminRequest = Request & { user: JwtAdminPayload };
