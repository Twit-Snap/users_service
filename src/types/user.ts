import { UserRegisterDto } from 'userAuth';

export interface User {
  id: number;
  username: string;
  email: string;
  name: string;
  lastname: string;
  birthdate: Date;
  createdAt: Date;
}

export type UserWithPassword = User & { password: string };

export type PublicUser = Omit<User, 'id' | 'email' | 'name' | 'lastname' >;

export interface IUserRepository {
  findByEmailOrUsername(emailOrUsername: string): Promise<UserWithPassword | null>;
  getList(): Promise<User[] | null>;
  get(id: number): Promise<User | null>;
  create(userData: UserRegisterDto): Promise<User>;
}
