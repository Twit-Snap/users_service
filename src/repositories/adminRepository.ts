import { Pool, QueryResult } from 'pg';
import {AdminInfoDto, Admin} from 'admin';
import {AlreadyExistError, InvalidCredentialsError} from '../types/customAdminErros';

class AdminRepository {
    private pool: Pool;

    constructor(pool: Pool) {
        this.pool = pool;
    }

    async create(adminData: AdminInfoDto): Promise<Admin> {

        const { username, email, password } = adminData;
        await this.checkUsername(username, email);
        await this.checkEmail(username, email);

        const query = `
      INSERT INTO admins (username, email, password)
      VALUES ($1, $2, $3)
      RETURNING username, email
    `;

        const result: QueryResult = await this.pool.query(query, [username, email, password]);
        return result.rows[0] as Admin;
    }

    async getAdminByEmail(email: string): Promise<AdminInfoDto> {

        const query = ` SELECT username, email, password FROM admins WHERE email = $1`;

        const result: QueryResult = await this.pool.query(query, [email]);
        if (result.rows.length === 0) {
            throw new InvalidCredentialsError("Invalid email");
        }
        return result.rows[0] as AdminInfoDto;
    }

    private async checkUsername(username: string, email: string) {
        const checkQuery = `SELECT username FROM admins WHERE username = $1`;
        const checkResult: QueryResult = await this.pool.query(checkQuery, [username]);

        const rowCount = checkResult.rowCount ?? 0;
        if (rowCount > 0) {
            throw new AlreadyExistError(username,email);
        }
    }

    private async checkEmail(username: string, email: string) {
        const checkQuery = `SELECT email FROM admins WHERE email = $1`;
        const checkResult: QueryResult = await this.pool.query(checkQuery, [email]);

        const rowCount = checkResult.rowCount ?? 0;
        if (rowCount > 0) {
            throw new AlreadyExistError(username,email);
        }
    }
}

export { AdminRepository };
