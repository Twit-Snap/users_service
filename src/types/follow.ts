import { TwitUser } from 'user';

export type FollowResponse = {
  username: string;
  followedUsername: string;
  createdAt: string;
};

export type FollowReturn = {
  createdAt: string;
};

export type FollowersResponse = {
  follower: Omit<TwitUser, 'id'>;
  createdAt: string;
};

export type FollowersReturn = {
  id: number;
  username: string;
  name: string;
  createdAt: string;
};
