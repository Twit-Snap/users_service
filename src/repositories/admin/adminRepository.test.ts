import { Pool } from 'pg';
import { AdminWithPassword } from '../../types/admin';
import { EntityAlreadyExistsError } from '../../types/customErrors';
import { AdminRepository } from './adminRepository';

function interpolateQuery(query: string, params: any[]): string {
  return query.replace(/\$\d+/g, (match) => {
    const index = parseInt(match.substr(1)) - 1;
    const value = params[index];
    return typeof value === 'string' ? `'${value}'` : value;
  });
}

describe('AdminRepository', () => {
  let adminRepository: AdminRepository;
  let mockPool: {
    query: jest.Mock;
  };

  beforeEach(() => {
    mockPool = {
      query: jest.fn()
    };
    adminRepository = new AdminRepository(mockPool as unknown as Pool);
  });

  describe('create', () => {
    const mockAdminData: AdminWithPassword = {
      username: 'testadmin',
      email: 'admin@test.com',
      password: 'hashedpassword123'
    };

    it('should create an admin successfully', async () => {
      mockPool.query.mockResolvedValueOnce({
        rows: [
          {
            username: mockAdminData.username,
            email: mockAdminData.email
          }
        ]
      });

      const result = await adminRepository.create(mockAdminData);

      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('create query and params');
      expect(result).toMatchSnapshot('created admin result');
      const [query, param] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, param);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should throw EntityAlreadyExistsError for duplicate username', async () => {
      const mockError = {
        code: '23505',
        constraint: 'admins_pkey'
      };

      mockPool.query.mockRejectedValueOnce(mockError);

      await expect(adminRepository.create(mockAdminData)).rejects.toThrow(EntityAlreadyExistsError);

      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('duplicate username query and params');
      const [query, param] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, param);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should throw EntityAlreadyExistsError for duplicate email', async () => {
      const mockError = {
        code: '23505',
        constraint: 'admins_email_key'
      };

      mockPool.query.mockRejectedValueOnce(mockError);

      await expect(adminRepository.create(mockAdminData)).rejects.toThrow(EntityAlreadyExistsError);

      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('duplicate email query and params');
      const [query, param] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, param);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should propagate unknown database errors', async () => {
      const mockError = new Error('Database connection failed');

      mockPool.query.mockRejectedValueOnce(mockError);

      await expect(adminRepository.create(mockAdminData)).rejects.toThrow(mockError);

      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('failed query and params');
      const [query, param] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, param);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });
  });

  describe('findByEmailOrUsername', () => {
    const mockAdmin: AdminWithPassword = {
      username: 'testadmin',
      email: 'admin@test.com',
      password: 'hashedpassword123'
    };

    it('should find admin by email', async () => {
      mockPool.query.mockResolvedValueOnce({
        rows: [mockAdmin]
      });

      const result = await adminRepository.findByEmailOrUsername('admin@test.com');

      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('find by email query and params');
      expect(result).toMatchSnapshot('found admin by email result');
      const [query, param] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, param);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should find admin by username', async () => {
      mockPool.query.mockResolvedValueOnce({
        rows: [mockAdmin]
      });

      const result = await adminRepository.findByEmailOrUsername('testadmin');

      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('find by username query and params');
      expect(result).toMatchSnapshot('found admin by username result');
      const [query, param] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, param);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should return null when admin not found', async () => {
      mockPool.query.mockResolvedValueOnce({
        rows: []
      });

      const result = await adminRepository.findByEmailOrUsername('nonexistent');

      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('not found query and params');
      expect(result).toMatchSnapshot('not found result');
      const [query, param] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, param);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should handle database error', async () => {
      const mockError = new Error('Database connection failed');

      mockPool.query.mockRejectedValueOnce(mockError);

      await expect(adminRepository.findByEmailOrUsername('testadmin')).rejects.toThrow(mockError);

      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('error query and params');
      const [query, param] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, param);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });
  });
});
