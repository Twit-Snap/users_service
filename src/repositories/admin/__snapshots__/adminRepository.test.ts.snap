// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminRepository create should create an admin successfully: Interpolated SQL query 1`] = `
"
      INSERT INTO admins (username, email, password)
      VALUES ('testadmin', 'admin@test.com', 'hashedpassword123')
      RETURNING username, email
    "
`;

exports[`AdminRepository create should create an admin successfully: create query and params 1`] = `
[
  "
      INSERT INTO admins (username, email, password)
      VALUES ($1, $2, $3)
      RETURNING username, email
    ",
  [
    "testadmin",
    "admin@test.com",
    "hashedpassword123",
  ],
]
`;

exports[`AdminRepository create should create an admin successfully: created admin result 1`] = `
{
  "email": "admin@test.com",
  "username": "testadmin",
}
`;

exports[`AdminRepository create should propagate unknown database errors: Interpolated SQL query 1`] = `
"
      INSERT INTO admins (username, email, password)
      VALUES ('testadmin', 'admin@test.com', 'hashedpassword123')
      RETURNING username, email
    "
`;

exports[`AdminRepository create should propagate unknown database errors: failed query and params 1`] = `
[
  "
      INSERT INTO admins (username, email, password)
      VALUES ($1, $2, $3)
      RETURNING username, email
    ",
  [
    "testadmin",
    "admin@test.com",
    "hashedpassword123",
  ],
]
`;

exports[`AdminRepository create should throw EntityAlreadyExistsError for duplicate email: Interpolated SQL query 1`] = `
"
      INSERT INTO admins (username, email, password)
      VALUES ('testadmin', 'admin@test.com', 'hashedpassword123')
      RETURNING username, email
    "
`;

exports[`AdminRepository create should throw EntityAlreadyExistsError for duplicate email: duplicate email query and params 1`] = `
[
  "
      INSERT INTO admins (username, email, password)
      VALUES ($1, $2, $3)
      RETURNING username, email
    ",
  [
    "testadmin",
    "admin@test.com",
    "hashedpassword123",
  ],
]
`;

exports[`AdminRepository create should throw EntityAlreadyExistsError for duplicate username: Interpolated SQL query 1`] = `
"
      INSERT INTO admins (username, email, password)
      VALUES ('testadmin', 'admin@test.com', 'hashedpassword123')
      RETURNING username, email
    "
`;

exports[`AdminRepository create should throw EntityAlreadyExistsError for duplicate username: duplicate username query and params 1`] = `
[
  "
      INSERT INTO admins (username, email, password)
      VALUES ($1, $2, $3)
      RETURNING username, email
    ",
  [
    "testadmin",
    "admin@test.com",
    "hashedpassword123",
  ],
]
`;

exports[`AdminRepository findByEmailOrUsername should find admin by email: Interpolated SQL query 1`] = `"SELECT username, email, password, created_at AS "createdAt" FROM admins WHERE email = 'admin@test.com' OR username = 'admin@test.com'"`;

exports[`AdminRepository findByEmailOrUsername should find admin by email: find by email query and params 1`] = `
[
  "SELECT username, email, password, created_at AS "createdAt" FROM admins WHERE email = $1 OR username = $1",
  [
    "admin@test.com",
  ],
]
`;

exports[`AdminRepository findByEmailOrUsername should find admin by email: found admin by email result 1`] = `
{
  "email": "admin@test.com",
  "password": "hashedpassword123",
  "username": "testadmin",
}
`;

exports[`AdminRepository findByEmailOrUsername should find admin by username: Interpolated SQL query 1`] = `"SELECT username, email, password, created_at AS "createdAt" FROM admins WHERE email = 'testadmin' OR username = 'testadmin'"`;

exports[`AdminRepository findByEmailOrUsername should find admin by username: find by username query and params 1`] = `
[
  "SELECT username, email, password, created_at AS "createdAt" FROM admins WHERE email = $1 OR username = $1",
  [
    "testadmin",
  ],
]
`;

exports[`AdminRepository findByEmailOrUsername should find admin by username: found admin by username result 1`] = `
{
  "email": "admin@test.com",
  "password": "hashedpassword123",
  "username": "testadmin",
}
`;

exports[`AdminRepository findByEmailOrUsername should handle database error: Interpolated SQL query 1`] = `"SELECT username, email, password, created_at AS "createdAt" FROM admins WHERE email = 'testadmin' OR username = 'testadmin'"`;

exports[`AdminRepository findByEmailOrUsername should handle database error: error query and params 1`] = `
[
  "SELECT username, email, password, created_at AS "createdAt" FROM admins WHERE email = $1 OR username = $1",
  [
    "testadmin",
  ],
]
`;

exports[`AdminRepository findByEmailOrUsername should return null when admin not found: Interpolated SQL query 1`] = `"SELECT username, email, password, created_at AS "createdAt" FROM admins WHERE email = 'nonexistent' OR username = 'nonexistent'"`;

exports[`AdminRepository findByEmailOrUsername should return null when admin not found: not found query and params 1`] = `
[
  "SELECT username, email, password, created_at AS "createdAt" FROM admins WHERE email = $1 OR username = $1",
  [
    "nonexistent",
  ],
]
`;

exports[`AdminRepository findByEmailOrUsername should return null when admin not found: not found result 1`] = `null`;
